From fc93cf60962cd2369eee895b8042cfd727dde8a7 Mon Sep 17 00:00:00 2001
From: Emergi Idan <idan.emergi@intel.com>
Date: Thu, 20 Feb 2020 11:41:01 +0200
Subject: [PATCH] PONRTSYS-7025:
 ppv4-reassembly-multitask-multicore-not-working

Add debugfs file for reading egress UC FW writer ports
configuration

Signed-off-by: Emergi Idan <idan.emergi@intel.com>
---
 drivers/net/ppv4/pp-hal/micro-controller/uc.c      | 109 +++++++++++++++++++--
 .../net/ppv4/pp-hal/micro-controller/uc_common.h   |   5 +-
 .../net/ppv4/pp-hal/micro-controller/uc_debugfs.c  |  57 ++++++++++-
 .../net/ppv4/pp-hal/micro-controller/uc_internal.h |  36 +++++++
 drivers/net/ppv4/pp-hal/micro-controller/uc_regs.h |  19 ++++
 .../pp-hal/micro-controller/uc_shared_engines.h    |   5 +
 6 files changed, 217 insertions(+), 14 deletions(-)

diff --git a/drivers/net/ppv4/pp-hal/micro-controller/uc.c b/drivers/net/ppv4/pp-hal/micro-controller/uc.c
index a71404590d59..8b4aa45add7c 100644
--- a/drivers/net/ppv4/pp-hal/micro-controller/uc.c
+++ b/drivers/net/ppv4/pp-hal/micro-controller/uc.c
@@ -175,10 +175,10 @@ bool uc_is_cpu_valid(bool uc_is_egr, u32 cid)
 /**
  * @brief Test whether a task id is valid or not
  * @param uc_is_egr select cluster
- * @param tid task if
+ * @param tid task id
  * @return bool true in case it is valid, false otherwise
  */
-static bool uc_is_task_is_valid(bool uc_is_egr, u32 tid)
+static bool uc_is_task_id_valid(bool uc_is_egr, u32 tid)
 {
 	u32 num_tasks = uc_is_egr ? UC_EGR_MAXTASK : UC_ING_MAXTASK;
 
@@ -190,6 +190,101 @@ static bool uc_is_task_is_valid(bool uc_is_egr, u32 tid)
 	return false;
 }
 
+/**
+ * @brief Test whether egress port id is valid or not
+ * @param port the port id
+ * @return bool true in case it is valid, false otherwise
+ */
+static bool uc_is_egr_port_valid(u32 port)
+{
+	if (port < UC_EGR_MAX_PORT)
+		return true;
+
+	pr_err("invalid egress port id %u\n", port);
+	return false;
+}
+
+s32 uc_egr_writer_port_en_get(u32 port, bool *en)
+{
+	u64 addr;
+
+	if (unlikely(ptr_is_null(en)))
+		return -EINVAL;
+
+	if (!uc_is_egr_port_valid(port))
+		return -EINVAL;
+
+	if (pp_silicon_step_get() == PP_SSTEP_A)
+		addr = UC_EGR_A_WRITER_PORT_ENABLE(port);
+	else
+		addr = UC_EGR_B_WRITER_PORT_ENABLE(port + UC_B_TXMGR_PORT_OFF);
+
+	*en = !!PP_REG_RD32(addr);
+
+	return 0;
+}
+
+s32 uc_egr_writer_map_get(u32 cid, u32 port, u32 *queue)
+{
+	u64 addr;
+
+	if (unlikely(ptr_is_null(queue)))
+		return -EINVAL;
+
+	if (!uc_is_cpu_valid(UC_IS_EGR, cid) || !uc_is_egr_port_valid(port))
+		return -EINVAL;
+
+	if (pp_silicon_step_get() == PP_SSTEP_A)
+		addr = UC_EGR_A_WRITER_MAP(cid, port);
+	else
+		addr = UC_EGR_A_WRITER_MAP(cid, port + UC_B_TXMGR_PORT_OFF);
+
+	*queue = PP_REG_RD32(addr);
+
+	return 0;
+}
+
+s32 uc_egr_writer_queue_credit_get(u32 cid, u32 port, u32 *credit)
+{
+	u64 addr;
+
+	if (unlikely(ptr_is_null(credit)))
+		return -EINVAL;
+
+	if (!uc_is_cpu_valid(UC_IS_EGR, cid) || !uc_is_egr_port_valid(port))
+		return -EINVAL;
+
+	if (pp_silicon_step_get() == PP_SSTEP_A)
+		addr = UC_EGR_A_WRITER_QUEUE_CREDIT(cid, port);
+	else
+		addr = UC_EGR_B_WRITER_QUEUE_CREDIT(cid,
+						    port + UC_B_TXMGR_PORT_OFF);
+
+	*credit = PP_REG_RD32(addr);
+
+	return 0;
+}
+
+s32 uc_egr_writer_port_credit_get(u32 port, u32 *credit)
+{
+	u64 addr;
+
+	if (unlikely(ptr_is_null(credit)))
+		return -EINVAL;
+
+	if (!uc_is_egr_port_valid(port))
+		return -EINVAL;
+
+	if (pp_silicon_step_get() == PP_SSTEP_A)
+		addr = UC_EGR_A_WRITER_PORT_CREDIT(port);
+	else
+		addr = UC_EGR_B_WRITER_PORT_CREDIT(port + UC_B_TXMGR_PORT_OFF);
+
+	*credit = PP_REG_RD32(addr);
+
+	return 0;
+}
+
 s32 uc_reass_stats_get(u32 cid, struct reassembly_stats *stats)
 {
 	u32 off;
@@ -216,7 +311,7 @@ s32 uc_task_pc_get(bool uc_is_egr, u32 cid, u32 tid, u32 *pc)
 	if (!uc_is_cpu_active(uc_is_egr, cid))
 		return -EINVAL;
 
-	if (!uc_is_task_is_valid(uc_is_egr, tid))
+	if (!uc_is_task_id_valid(uc_is_egr, tid))
 		return -EINVAL;
 
 	if (pp_silicon_step_get() == PP_SSTEP_A) {
@@ -246,7 +341,7 @@ s32 uc_task_status32_get(bool uc_is_egr, u32 cid, u32 tid, u32 *status)
 	if (!uc_is_cpu_active(uc_is_egr, cid))
 		return -EINVAL;
 
-	if (!uc_is_task_is_valid(uc_is_egr, tid))
+	if (!uc_is_task_id_valid(uc_is_egr, tid))
 		return -EINVAL;
 
 	if (pp_silicon_step_get() == PP_SSTEP_A) {
@@ -275,7 +370,7 @@ s32 uc_task_ready_get(bool uc_is_egr, u32 cid, u32 tid, bool *ready)
 	if (!uc_is_cpu_active(uc_is_egr, cid))
 		return -EINVAL;
 
-	if (!uc_is_task_is_valid(uc_is_egr, tid))
+	if (!uc_is_task_id_valid(uc_is_egr, tid))
 		return -EINVAL;
 
 	if (uc_is_egr)
@@ -301,7 +396,7 @@ s32 uc_task_popa_active_get(bool uc_is_egr, u32 cid, u32 tid, bool *active,
 	if (!uc_is_cpu_active(uc_is_egr, cid))
 		return -EINVAL;
 
-	if (!uc_is_task_is_valid(uc_is_egr, tid))
+	if (!uc_is_task_id_valid(uc_is_egr, tid))
 		return -EINVAL;
 
 	if (uc_is_egr)
@@ -352,7 +447,7 @@ s32 uc_task_regs_get(bool uc_is_egr, u32 cid, u32 tid, u32 *regs, u32 num)
 	if (!uc_is_cpu_active(uc_is_egr, cid))
 		return -EINVAL;
 
-	if (!uc_is_task_is_valid(uc_is_egr, tid))
+	if (!uc_is_task_id_valid(uc_is_egr, tid))
 		return -EINVAL;
 
 	for (i = 0; i < min(num, UC_TASK_REGS_NUM); i++) {
diff --git a/drivers/net/ppv4/pp-hal/micro-controller/uc_common.h b/drivers/net/ppv4/pp-hal/micro-controller/uc_common.h
index f61735ff51ac..3eead3525744 100644
--- a/drivers/net/ppv4/pp-hal/micro-controller/uc_common.h
+++ b/drivers/net/ppv4/pp-hal/micro-controller/uc_common.h
@@ -13,7 +13,7 @@
  * the version here
  */
 #define EGRESS_VER_MAJOR  1
-#define EGRESS_VER_MINOR  11
+#define EGRESS_VER_MINOR  12
 
 #define INGRESS_VER_MAJOR 1
 #define INGRESS_VER_MINOR 2
@@ -111,7 +111,8 @@ struct uc_log_msg {
 	u32  val;
 	u32  ts;
 	u8   level;
-	u8   rsrv;
+	u8   cid:3,
+	     tid:5;
 	char str[38];
 	char func[16];
 };
diff --git a/drivers/net/ppv4/pp-hal/micro-controller/uc_debugfs.c b/drivers/net/ppv4/pp-hal/micro-controller/uc_debugfs.c
index ae688bbd5f41..361cdf46119d 100644
--- a/drivers/net/ppv4/pp-hal/micro-controller/uc_debugfs.c
+++ b/drivers/net/ppv4/pp-hal/micro-controller/uc_debugfs.c
@@ -466,9 +466,9 @@ static void uc_dbg_log_show(struct seq_file *f)
 	for_each_arr_entry(msg, msgs, sz / sizeof(*msg)) {
 		if (strlen(msg->str) == 0)
 			continue;
-		seq_printf(f, "[%8x]:[%-5s]:[%s]: %s %u, %#x\n",
-			   msg->ts, level_str[msg->level], msg->func,
-			   msg->str, msg->val, msg->val);
+		seq_printf(f, "[%8x]:[%-5s]:[%u.%02u][%s]: %s %u, %#x\n",
+			   msg->ts, level_str[msg->level], msg->cid, msg->tid,
+			   msg->func, msg->str, msg->val, msg->val);
 	}
 	seq_puts(f, "\n");
 
@@ -775,16 +775,62 @@ static void uc_task_show(bool uc_is_egr, struct seq_file *f)
 
 static void uc_egr_task_show(struct seq_file *f)
 {
-	uc_task_show(true, f);
+	uc_task_show(UC_IS_EGR, f);
 }
 PP_DEFINE_DEBUGFS(egr_task, uc_egr_task_show, uc_task_set);
 
 static void uc_ing_task_show(struct seq_file *f)
 {
-	uc_task_show(true, f);
+	uc_task_show(UC_IS_ING, f);
 }
 PP_DEFINE_DEBUGFS(ing_task, uc_ing_task_show, uc_task_set);
 
+static void uc_egr_writer_show(struct seq_file *f)
+{
+	u32 port, cid, queue, credits;
+	bool en;
+	u8 num_cpu;
+	static const char *const port_str[UC_EGR_MAX_PORT] = {
+		"MULTICAST", "REASSEMBLY", "FRAGMENTATION",
+		"TURBODOX",  "IPSEC",
+	};
+
+	uc_ccu_maxcpus_get(UC_IS_EGR, &num_cpu);
+
+	seq_puts(f,
+		 " +---------------------------------------------------------------+\n");
+	seq_printf(f, " | %-16s", "Port");
+	seq_printf(f, " | %-7s", "Enable");
+	seq_printf(f, " | %-6s", "Credit");
+	seq_printf(f, " | %-8s", "Mapping");
+	seq_printf(f, " | %12s |\n", "Queue Credit");
+	seq_puts(f,
+		 " +------------------+---------+--------+----------+--------------+\n");
+
+	for (port = 0; port < UC_EGR_MAX_PORT; port++) {
+		uc_egr_writer_port_en_get(port, &en);
+		uc_egr_writer_port_credit_get(port, &credits);
+		seq_puts(f, " |");
+		seq_printf(f, " %-2u-%-13s |", port, port_str[port]);
+		seq_printf(f, " %-7s |", BOOL2EN(en));
+		seq_printf(f, " %-6u |", credits);
+
+		for (cid = 0; cid < (u32)num_cpu; cid++) {
+			if (cid != 0)
+				seq_puts(f,
+					 " |                  |         |        |");
+			uc_egr_writer_map_get(cid, port, &queue);
+			uc_egr_writer_queue_credit_get(cid, port, &credits);
+			seq_printf(f, " CPU%u: Q%u |", cid, queue);
+			seq_printf(f, " %-12u |\n", credits);
+		}
+		seq_puts(f,
+			 " |------------------+---------+--------+----------+--------------|\n");
+	}
+}
+
+PP_DEFINE_DEBUGFS(egr_writer, uc_egr_writer_show, NULL);
+
 static struct debugfs_file uc_dbg_files[] = {
 	{"nf_cfg", &PP_DEBUGFS_FOPS(nf_cfg)},
 	{"core_run_halt", &PP_DEBUGFS_FOPS(core_run_halt)},
@@ -800,6 +846,7 @@ static struct debugfs_file uc_dbg_files[] = {
 	{"egr_task", &PP_DEBUGFS_FOPS(egr_task)},
 	{"ing_tasks", &PP_DEBUGFS_FOPS(ing_tasks)},
 	{"ing_task", &PP_DEBUGFS_FOPS(ing_task)},
+	{"egr_writer", &PP_DEBUGFS_FOPS(egr_writer)},
 };
 
 s32 uc_dbg_init(struct dentry *parent)
diff --git a/drivers/net/ppv4/pp-hal/micro-controller/uc_internal.h b/drivers/net/ppv4/pp-hal/micro-controller/uc_internal.h
index 777329ba6822..579ebee837b3 100644
--- a/drivers/net/ppv4/pp-hal/micro-controller/uc_internal.h
+++ b/drivers/net/ppv4/pp-hal/micro-controller/uc_internal.h
@@ -59,6 +59,42 @@ s32 uc_egr_mbox_init(void);
  */
 void uc_egr_mbox_exit(void);
 
+/**
+ * @brief Get egress writter port enable
+ * @param port port id
+ * @param en
+ * @return s32 0 on success, error code otherwise
+ */
+s32 uc_egr_writer_port_en_get(u32 port, bool *en);
+
+/**
+ * @brief Get the writer port and queue mapping
+ *        for a specific CPU
+ * @param cid cpu id
+ * @param port port id
+ * @param queue the mapped queue
+ * @return s32 0 on success, error code otherwise
+ */
+s32 uc_egr_writer_map_get(u32 cid, u32 port, u32 *queue);
+
+/**
+ * @brief Get the writer credits for a the queue
+ *        which is mapped to 'port' on CPU 'cid'
+ * @param cid cpu id
+ * @param port port id
+ * @param credit the credits count
+ * @return s32 0 on success, error code otherwise
+ */
+s32 uc_egr_writer_queue_credit_get(u32 cid, u32 port, u32 *credit);
+
+/**
+ * @brief Get the writer credits for a specific port
+ * @param port port id
+ * @param credit the credits count
+ * @return s32 0 on success, error code otherwise
+ */
+s32 uc_egr_writer_port_credit_get(u32 port, u32 *credit);
+
 /**
  * @brief Get the program counter register of CPU's task
  * @param uc_is_egr select the uc cluster
diff --git a/drivers/net/ppv4/pp-hal/micro-controller/uc_regs.h b/drivers/net/ppv4/pp-hal/micro-controller/uc_regs.h
index eec5f44f0f0d..be3d64ed82fc 100644
--- a/drivers/net/ppv4/pp-hal/micro-controller/uc_regs.h
+++ b/drivers/net/ppv4/pp-hal/micro-controller/uc_regs.h
@@ -315,6 +315,25 @@ extern u64 uc_ing_base_addr;
 #define UC_EGR_B_WRITER_PORT_ADDR(p)   \
 	UC_B_WRITER_PORT_SPEC(EGRESS, p, 4)
 
+/* WRITER_PORT MAP */
+#define UC_EGR_A_WRITER_MAP(c, p) \
+	(UC_SHRD_REG_IDX(EGRESS, WRITER_A, MAP, p) | (c << 6))
+#define UC_EGR_B_WRITER_MAP(c, p) \
+	(UC_SHRD_REG_IDX(EGRESS, WRITER_B, MAP, p) | (c << 6))
+
+/* WRITER_PORT QUEUE CREDIT */
+#define UC_EGR_A_WRITER_QUEUE_CREDIT(c, p) \
+	(UC_SHRD_REG_IDX(EGRESS, WRITER_A, QUEUE_CREDIT, p) | (c << 6))
+#define UC_EGR_B_WRITER_QUEUE_CREDIT(c, p) \
+	(UC_SHRD_REG_IDX(EGRESS, WRITER_B, QUEUE_CREDIT, p) | (c << 6))
+
+/* WRITER_PORT_CREDIT */
+#define UC_EGR_A_WRITER_PORT_CREDIT(p) \
+	UC_SHRD_REG_IDX(EGRESS, WRITER_A, PORT_CREDIT, p)
+#define UC_EGR_B_WRITER_PORT_CREDIT(p)   \
+	UC_B_WRITER_PORT_SPEC(EGRESS, p, 2)
+
+
 /***** WRITER MB EGRESS REGISTERS *****/
 /* WRITER_MB_MESSAGE */
 #define UC_EGR_MBOX_MSG0_WR_ADDRESS    \
diff --git a/drivers/net/ppv4/pp-hal/micro-controller/uc_shared_engines.h b/drivers/net/ppv4/pp-hal/micro-controller/uc_shared_engines.h
index da8751d39fdf..e5caa326bc3e 100644
--- a/drivers/net/ppv4/pp-hal/micro-controller/uc_shared_engines.h
+++ b/drivers/net/ppv4/pp-hal/micro-controller/uc_shared_engines.h
@@ -52,6 +52,11 @@
 #define WRITER_B_PORT_ENABLE    16
 
 #define WRITER_A_PORT           10
+#define WRITER_A_MAP            13
+#define WRITER_B_MAP            1
+#define WRITER_A_QUEUE_CREDIT   14
+#define WRITER_B_QUEUE_CREDIT   2
+#define WRITER_A_PORT_CREDIT    16
 #define WRITER_A_PORT_DEPTH     20
 
 /* Writer_MBOX Egress: */
